File: scripts/world/Growth.lua
-- ============================================================================
-- Growth — 城市成长引擎（都市天际线核心循环）
--   分区内、邻路的空地，按 RCI 需求自动长出建筑；
--   已有成长楼在需求高 + 地价高（服务覆盖）时升级 1→3 级
-- ============================================================================

local Config = require("Config")
local World = require("world.World")

local Growth = {}

local G = Config.GROWTH

local acc_ = 0

Growth.simTime = 0          -- 累计模拟时间（秒，含速度倍率），生长动画用
Growth.lastDemand = { r = 0.5, c = 0.3, i = 0.4 }
Growth.grewCount = 0        -- 本次 tick 长了几栋（UI 提示/音效用）

function Growth.reset()
  acc_ = 0
  Growth.simTime = 0
  Growth.grewCount = 0
end

-- RCI 需求（0..1），由 GameData 传入统计
function Growth.computeDemand(st, pop)
  local normR = math.max(st.resCap + 40, 1)
  local normC = math.max(st.comCap + 30, 1)
  local normI = math.max(st.indCap + 40, 1)
  local r = (st.comCap + st.indCap) * 1.15 + 40 - pop      -- 岗位缺口 → 住宅需求
  local c = pop * 0.55 - st.comCap                          -- 消费缺口 → 商业需求
  local i = pop * 0.45 + 50 - st.indCap                     -- 外部订单 → 工业需求
  return {
    r = math.max(0, math.min(1, r / normR)),
    c = math.max(0, math.min(1, c / normC)),
    i = math.max(0, math.min(1, i / normI)),
  }
end

-- 一次成长判定
local function growthStep(sfx)
  local w = World.current
  if not w then return end
  local st = World.stats()
  local pop = World._pop or 0
  local demand = Growth.computeDemand(st, pop)
  Growth.lastDemand = demand

  -- 按需求权重选本步要生长的类型
  local pool = {}
  if demand.r >= G.demandMin then pool[#pool + 1] = { z = "residential", w = demand.r } end
  if demand.c >= G.demandMin then pool[#pool + 1] = { z = "commercial", w = demand.c } end
  if demand.i >= G.demandMin then pool[#pool + 1] = { z = "industrial", w = demand.i } end
  Growth.grewCount = 0
  if #pool == 0 then return end

  -- 收集候选空地：分区匹配 + 邻路
  local want = {}
  for _, p in ipairs(pool) do want[p.z] = true end
  local candidates = {}
  for y = 2, w.rows - 1 do
    local row = w.grid[y]
    for x = 2, w.cols - 1 do
      local t = row[x]
      if want[t.zone] and not t.building and not t.road and t.terrain ~= "water" then
        if World.isRoad(x + 1, y) or World.isRoad(x - 1, y)
        or World.isRoad(x, y + 1) or World.isRoad(x, y - 1) then
          candidates[#candidates + 1] = { x = x, y = y, zone = t.zone }
        end
      end
    end
  end
  if #candidates > 0 and math.random() < G.spawnChance then
    local pick = candidates[math.random(#candidates)]
    if World.growBuilding(pick.zone, pick.x, pick.y, 1, Growth.simTime) then
      Growth.grewCount = Growth.grewCount + 1
      if sfx then sfx("sfx_build", 0.35) end
    end
  end

  -- 升级判定：需求高 + 概率；服务覆盖（地价）提高概率
  local upCandidates = {}
  for _, e in ipairs(World.allBuildings()) do
    local b = e.b
    if not b.service and b.level < #Config.GROWN[b.zone].levels then
      upCandidates[#upCandidates + 1] = e
    end
  end
  if #upCandidates > 0 and math.random() < G.upgradeChance then
    local e = upCandidates[math.random(#upCandidates)]
    local zoneDemand = e.b.zone == "residential" and demand.r
                    or (e.b.zone == "commercial" and demand.c or demand.i)
    if zoneDemand >= G.demandMin and math.random() < zoneDemand then
      World.upgradeBuilding(e.x, e.y)
    end
  end
end

---@param dt number 已含速度倍率的模拟时间（秒）
---@param sfx fun(name: string, gain: number)|nil
function Growth.tick(dt, sfx)
  Growth.simTime = Growth.simTime + dt
  acc_ = acc_ + dt
  if acc_ >= G.tickSeconds then
    acc_ = acc_ - G.tickSeconds
    growthStep(sfx)
  end
end

return Growth

File: android/app/src/main/java/com/chenguang/skylines/world/Growth.kt
package com.chenguang.skylines.world

import com.chenguang.skylines.Config
import com.chenguang.skylines.GameData
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

// ============================================================================
// Growth — 城市成长引擎，与 scripts/world/Growth.lua 1:1 对应
//   分区内、邻路的空地，按 RCI 需求自动长出建筑；
//   已有成长楼在需求高 + 地价高（服务覆盖）时升级 1→3 级
// ============================================================================

object Growth {

    private val G = Config.GROWTH

    private var acc: Double = 0.0

    /** 累计模拟时间（秒，含速度倍率），生长动画用 */
    var simTime: Double = 0.0

    data class Demand(var r: Double, var c: Double, var i: Double)

    var lastDemand: Demand = Demand(0.5, 0.3, 0.4)

    /** 本次 tick 长了几栋 */
    var grewCount: Int = 0

    fun reset() {
        acc = 0.0
        simTime = 0.0
        grewCount = 0
        lastDemand = Demand(0.5, 0.3, 0.4)
    }

    fun computeDemand(st: WorldStats, pop: Int): Demand {
        val s = GameData.current
        val edu = s?.education ?: 18.0
        val happy = s?.happiness ?: 60.0
        val taxRes = s?.taxRes ?: Config.TAX.default
        val taxCom = s?.taxCom ?: Config.TAX.default
        val taxInd = s?.taxInd ?: Config.TAX.default
        val taxDragR = 1.0 - max(0, taxRes - Config.TAX.default) * 0.04
        val taxDragC = 1.0 - max(0, taxCom - Config.TAX.default) * 0.04
        val taxDragI = 1.0 - max(0, taxInd - Config.TAX.default) * 0.04
        val overzoneR = max(0, st.resCount - 8) * 1.6
        val overzoneC = max(0, st.comCount - 6) * 1.4
        val overzoneI = max(0, st.indCount - 6) * 1.4
        val attract = (happy - 50) * 0.35 + (edu - 20) * 0.12
        val r = ((st.comCap + st.indCap) * 1.15 + 40 + attract - pop - overzoneR) * taxDragR
        val c = (pop * (0.50 + edu / 400.0) + 25 - st.comCap - overzoneC) * taxDragC
        val i = (pop * 0.45 + 50 + edu * 0.2 - st.indCap - overzoneI) * taxDragI
        val normR = max(st.resCap + 40.0, 1.0)
        val normC = max(st.comCap + 30.0, 1.0)
        val normI = max(st.indCap + 40.0, 1.0)
        return Demand(
            r = max(0.0, min(1.0, r / normR * GameData.policyMul("demandR"))),
            c = max(0.0, min(1.0, c / normC * GameData.policyMul("demandC"))),
            i = max(0.0, min(1.0, i / normI * GameData.policyMul("demandI")))
        )
    }

    private fun growthStep(sfx: ((String, Float) -> Unit)?) {
        val w = World.current ?: return
        val st = World.stats()
        val pop = w._pop
        val demand = computeDemand(st, pop)
        lastDemand = demand

        // 按需求权重选本步要生长的类型
        val pool = mutableListOf<Pair<String, Double>>()
        if (demand.r >= G.demandMin) pool.add("residential" to demand.r)
        if (demand.c >= G.demandMin) pool.add("commercial" to demand.c)
        if (demand.i >= G.demandMin) pool.add("industrial" to demand.i)

        grewCount = 0
        if (pool.isEmpty()) return

        // 收集候选空地：分区匹配 + 邻路
        val want = pool.map { it.first }.toSet()
        val candidates = mutableListOf<Triple<Int, Int, String>>()
        for (y in 2 until w.rows) {
            for (x in 2 until w.cols) {
                val t = w.grid[y - 1][x - 1]
                if (t.zone in want && t.building == null && t.road == null && t.terrain != "water") {
                    if (World.isRoad(x + 1, y) || World.isRoad(x - 1, y) ||
                        World.isRoad(x, y + 1) || World.isRoad(x, y - 1)
                    ) {
                        candidates.add(Triple(x, y, t.zone))
                    }
                }
            }
        }
        if (candidates.isNotEmpty() && Random.nextDouble() < G.spawnChance) {
            val pick = candidates[Random.nextInt(candidates.size)]
            if (World.growBuilding(pick.third, pick.first, pick.second, 1, simTime)) {
                grewCount++
                sfx?.invoke("sfx_build", 0.35f)
            }
        }

        // 升级判定
        val upCandidates = World.allBuildings().filter {
            !it.b.isService && it.b.level < (Config.GROWN[it.b.zone]?.levels?.size ?: 1)
        }
        val upgradeChance = G.upgradeChance * GameData.policyMul("upgradeMul")
        if (upCandidates.isNotEmpty() && Random.nextDouble() < upgradeChance) {
            val e = upCandidates[Random.nextInt(upCandidates.size)]
            val zoneDemand = when (e.b.zone) {
                "residential" -> demand.r
                "commercial" -> demand.c
                else -> demand.i
            }
            val ageOk = (simTime - e.b.born) >= G.upgradeAgeDays * Config.TIME.daySeconds
            val land = World.landValue(e.x, e.y)
            val covOk = World.isCoveredBy(e.x, e.y, Config.ServiceCat.POWER) &&
                World.isCoveredBy(e.x, e.y, Config.ServiceCat.WATER)
            if (zoneDemand >= G.demandMin && ageOk && land >= G.landValueUpgrade && covOk &&
                Random.nextDouble() < zoneDemand
            ) {
                World.upgradeBuilding(e.x, e.y)
            }
        }
    }

    /** dt：已含速度倍率的模拟时间（秒） */
    fun tick(dt: Double, sfx: ((String, Float) -> Unit)?) {
        simTime += dt
        acc += dt
        if (acc >= G.tickSeconds) {
            acc -= G.tickSeconds
            growthStep(sfx)
        }
    }
}
